src/commons/authentication/mod.rs

```diff
@@ -38,31 +37,106 @@ pub struct AuthenticationClassSpec {
             .await
     }
 }
 
+/// Common [`ClientAuthenticationDetails`] which is specified at the client/
+/// product cluster level. It provides a name (key) to resolve a particular
+/// [`AuthenticationClass`]. Additionally, it provides authentication provider
+/// specific configuration (OIDC and LDAP for example).
+///
+/// If the product needs additional (product specific) authentication options,
+/// it is recommended to wrap this struct and use `#[serde(flatten)]` on the
+/// field.
+///
+/// Additionally, it might be the case that special fields are needed in the
+/// contained structs, such as [`oidc::ClientAuthenticationOptions`]. To be able
+/// to add custom fields in that structs without serde(flattening) multiple structs,
+/// they are generic, so you can add additional attributes if needed.
+///
+/// ### Example
+///
+/// ```
+/// # use schemars::JsonSchema;
+/// # use serde::{Deserialize, Serialize};
+/// use stackable_operator::commons::authentication::ClientAuthenticationDetails;
+///
+/// #[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+/// #[serde(rename_all = "camelCase")]
+/// pub struct SupersetAuthenticationClass {
+///     pub user_registration_role: String,
+///     pub user_registration: bool,
+///
+///     #[serde(flatten)]
+///     pub common: ClientAuthenticationDetails,
+/// }
+/// ```
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[schemars(description = "")]
+pub struct ClientAuthenticationDetails<O = ()> {
+    /// A name/key which references an authentication class. To get the concrete
+    /// [`AuthenticationClass`], we must resolve it. This resolution can be
+    /// achieved by using [`ClientAuthenticationDetails::resolve_class`].
+    #[serde(rename = "authenticationClass")]
+    authentication_class_ref: String,
+
+    /// This field contains authentication provider specific configuration. It
+    /// is flattened into the final CRD.
+    ///
+    /// Use [`ClientAuthenticationDetails::oidc_or_error`] to get the value or report an error to the user.
+    oidc: Option<oidc::ClientAuthenticationOptions<O>>,
+}
+
+impl<O> ClientAuthenticationDetails<O> {
+    /// Resolves this specific [`AuthenticationClass`]. Usually products support
+    /// a list of authentication classes, which individually need to be resolved.
+    pub async fn resolve_class(&self, client: &Client) -> OperatorResult<AuthenticationClass> {
+        AuthenticationClass::resolve(client, &self.authentication_class_ref).await
+    }
+
+    pub fn authentication_class_name(&self) -> &String {
+        &self.authentication_class_ref
+    }
+
+    /// In case OIDC is configured, the user *needs* to provide some connection details,
+    /// such as the client credentials. Call this function in case the user has configured
+    /// OIDC, as it will error out then the OIDC client details are missing.
+    pub fn oidc_or_error(
```

Comment:
If we provide these kind of functions (as its super hard / impossible to solve this via types), we should do it for all methods, e.g. LDAP. But we might want to do this in a separate PR.

Reply:
We will add such as function for ldap once we switched that to new new structure as well